mono/mini/mini-amd64.c

```diff
@@ -4573,23 +4573,41 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
 				amd64_sse_movsd_reg_reg (code, ins->dreg, ins->sreg1);
 			break;
 		}
-		case OP_TAILCALL: {
+		case OP_TAILCALL:
+		case OP_TAILCALL_MEMBASE: {
 			MonoCallInst *call = (MonoCallInst*)ins;
 			int i, save_area_offset;
+			gboolean membase = (ins->opcode == OP_TAILCALL_MEMBASE);
+
+			// FIXME make this two pass for an accurate size
+			// FIXME varargs
 
 			g_assert (!cfg->method->save_lmf);
 
 			/* the size of the tailcall op depends on signature, let's check for enough
 			 * space in the code buffer here again */
 			max_len += AMD64_NREG * 4 + call->stack_usage * 15 + EXTRA_CODE_SPACE;
 
+			if (membase)
+				max_len += 3;
+
 			if (G_UNLIKELY (offset + max_len > cfg->code_size)) {
 				cfg->code_size *= 2;
 				cfg->native_code = (unsigned char *) mono_realloc_native_code(cfg);
 				code = cfg->native_code + offset;
 				cfg->stat_code_reallocs++;
 			}
 
+			// FIXME This is overly pessimistic.
+			// Rax can work, swap with below.
+			// R10 can work?
+			// Unused parameter registers are also ok.
+			// non-volatiles cannot work.
+			if (membase) {
+				amd64_mov_reg_reg (code, AMD64_R11, ins->sreg1, 8);
+				ins->sreg1 = AMD64_R11;
+			}
```

Comment:
This should use a local and not change ins->sreg1.

Reply:
This is based on CALL_REG. Originally I was confused and thought I needed TAILCALL_REG instead of TAILCALL_MEMBASE, but CALL_REG might still be relevant. Question is: is the data read again? If so, changing would seem maybe needed? CALL_REG specifically has to vacate RAX for vargs (at least on sysv, it does it unnecessarily on NT). Or the data is important to later analysis and we should insert a mov ins? I don't think there is later analysis, we are outputting bytes and no more decisions can be had.

Reply:
It just feels better not to rewrite it, just to be safe.

Reply:
Seems like maybe it is a dead store and remove it? I guess my favorate is remove it, or if we need accurate data, insert another ins?? Not sure ins buiding/changing is reasonable here, but CALL_REG does change it like this.

Reply:
It feels kinda safe either way for different reasons -- either it is read later and needs the change, or it was already read and it can't change, or some other. I think it isn't needed and let's try without.
CALL_REG could maybe use cleanup then. I was not certain of this line. i was following a pattern, I thought, but that isn't clear now. Maybe in the other case it really can/should be a local. They read it again.